### PR TITLE
Intrusions

### PR DESCRIPTION
**Pull request by lachlangrose**

### CODE REVIEW

**Comment by lachlangrose** [comment]
initial review

File: LoopStructural/modelling/core/geological_model.py [hunk 1]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
...
+        )
+
+        INet.set_data()
+        INet.set_contact_anisotropies(contacts_anisotropies, **kwargs)
+        INet.set_faults_anisotropies(structures_anisotropies)
+        INet.set_sequence_of_exploited_anisotropies(sequence_anisotropies)
+        INet.set_velocity_parameters()
+        INet.set_sections_axis(inet_axis)
+        logger.info("building intrusion network")
+        INet.build(**kwargs)
```
Comment (lachlangrose) [inline LoopStructural/modelling/core/geological_model.py:1130]: Can we avoid running "build" within the create_and_add function? I have tried to move all of the interpolation to be run on demand when the geological feature is evaluated or when builder.update is called

File: LoopStructural/modelling/core/geological_model.py [hunk 2]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
...
+        # Create intrusion frame, using intrusion network points, propagation and inflation direction
+
+        gxxgz = 0  # weight for orthogonality constraint between coord 0 and coord 2
+        gxxgy = 0  # weight for orthogonality constraint between coord 0 and coord 1
+        gyxgz = 0  # weight for orthogonality constraint between coord 1 and coord 2
+
+        weights = [gxxgz, gxxgy, gyxgz]
+        # reg=np.array([1,0.5,1])
+        logger.info("building intrusion frame")
+        interpolator = self.get_interpolator(interpolatortype="FDI")
```
Comment (lachlangrose) [inline LoopStructural/modelling/core/geological_model.py:1141]: avoid hard code

File: LoopStructural/modelling/core/geological_model.py [hunk 3]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
...
+        frame_data = self.data[self.data["feature_name"] == intrusion_frame_name].copy()
+        IFrame_builder = IntrusionBuilder(
+            interpolator, model=self, feature_name=intrusion_frame_name
+        )
+        IFrame_builder.set_data(frame_data, INet.intrusion_network_outcome)
+        IFrame = IFrame_builder.build(
```
Comment (lachlangrose) [inline LoopStructural/modelling/core/geological_model.py:1147]: don't call build during create_and_add

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 4]
```diff
@@ -0,0 +1,1118 @@
...
+from LoopStructural.utils import getLogger
+
+logger = getLogger(__name__)
+# GSLIB library
+try:
+    import geostatspy.GSLIB as GSLIB  # GSLIB utilities, viz and wrapped functions
+    import geostatspy.geostats as geostats  # GSLIB converted to Python
+
+except ImportError:
+    logger.error("GeostatPy not installed \n" "pip install geostatspy")
```
[the remark this answers is not in the document]
Comment (lachlangrose) [reply LoopStructural/modelling/intrusions/intrusion_body.py:13]: don't worry about this i see its covered in the create_and_add function with the boolean flag

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 5]
```diff
@@ -0,0 +1,1118 @@
...
+        """Set data for vertical extent (distances in G axis) simulation
+        Evaluate data points in the intrusion frame coodinates,
+        and separate data between intrusion network contact data and other contact data.
+
+        Parameters
+        ----------
+
+        Returns
+        ----------
+        """
```
Comment (lachlangrose) [inline LoopStructural/modelling/intrusions/intrusion_body.py:152]: can you add some comments to say what this all does. I am a bit lost

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 6]
```diff
@@ -0,0 +1,1118 @@
...
+            ndmax = lateral_simulation_parameters["ndmax"]
+        else:
+            ndmax = 3
+
+        if "radius" in lateral_simulation_parameters:
+            radius = lateral_simulation_parameters["radius"]
+        else:
+            radius = 500
+
+        s_parameters = {
```
Comment (lachlangrose) [inline LoopStructural/modelling/intrusions/intrusion_body.py:340]: You can probably simplify all of this by replacing the tmin,tmax variables with the kwarg.get('key',default_val) for constructing this dictionary

File: LoopStructural/modelling/core/geological_model.py [hunk 7]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
...
+        INet.set_faults_anisotropies(structures_anisotropies)
+        INet.set_sequence_of_exploited_anisotropies(sequence_anisotropies)
+        INet.set_velocity_parameters()
+        INet.set_sections_axis(inet_axis)
+        logger.info("building intrusion network")
+        INet.build(**kwargs)
+
+        # Create intrusion frame, using intrusion network points, propagation and inflation direction
+
+        gxxgz = 0  # weight for orthogonality constraint between coord 0 and coord 2
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/core/geological_model.py:1134]: done

File: LoopStructural/modelling/core/geological_model.py [hunk 8]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
...
+        logger.info("building intrusion frame")
+        interpolator = self.get_interpolator(interpolatortype="FDI")
+        frame_data = self.data[self.data["feature_name"] == intrusion_frame_name].copy()
+        IFrame_builder = IntrusionBuilder(
+            interpolator, model=self, feature_name=intrusion_frame_name
+        )
+        IFrame_builder.set_data(frame_data, INet.intrusion_network_outcome)
+        IFrame = IFrame_builder.build(
+            nelements=1e2,
+            solver="lu",
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/core/geological_model.py:1149]: done

File: LoopStructural/modelling/core/geological_model.py (hunk 2 above)
Comment (Fer071989) [reply LoopStructural/modelling/core/geological_model.py:1141]: done

File: LoopStructural/modelling/intrusions/geom_conceptual_models.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    #     b = abs(minS)
+
+    po = minP + (maxP - minP) / 2
+
+    p_locations = lateral_contact_data.loc[:, "coord1"].copy().to_numpy()
+    #     s_values = lateral_contact_data.loc[:,'coord2'].copy().to_numpy()
+
+    s = np.zeros([len(p_locations), 2])
+
+    for i in range(len(p_locations)):
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/geom_conceptual_models.py:31]: done

File: LoopStructural/modelling/intrusions/geom_conceptual_models.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    po = minP + a  # p coordinate of ellipsis centre
+    so = minS + b  # s coordinate of ellipsis centre
+
+    alpha = vertex[0]  # p coordinate of vertex
+    beta = vertex[1]  # g coordinate of vertex
+    gamma = vertex[2]  # s coordinate of vertex
+
+    growth = np.zeros([len(ps_locations), 2])  # container for results
+
+    for i in range(len(ps_locations)):
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/geom_conceptual_models.py:119]: done

File: LoopStructural/modelling/intrusions/geom_conceptual_models.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    s = np.zeros([len(p_locations), 2])
+
+    for i in range(len(p_locations)):
+        if minP < p_locations[i] < maxP:
+            s[i, 0] = maxS  # max side
+            s[i, 1] = minS  # min side
+        else:
+            s[i, :] = 0
+
+    return s
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/geom_conceptual_models.py:67]: done

File: LoopStructural/modelling/core/geological_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/core/geological_model.py:1130]: yes, these points are constraints for coordinate 0 of structural frame.

File: LoopStructural/modelling/core/geological_model.py [hunk 12]
```diff
@@ -1046,7 +1054,157 @@ def create_and_add_folded_fold_frame(
 
         self._add_feature(folded_fold_frame)
 
-        return folded_fold_frame
+        return fold_frame
+
+    def create_and_add_intrusion(
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/core/geological_model.py:1059]: Added to description of create_and_add_intrusion() function

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 13]
```diff
@@ -0,0 +1,1118 @@
...
+        Evaluate data points in the intrusion frame coodinates,
+        and separates data between sides (> or <0)
+
+        Parameters
+        ----------
+
+        Returns
+        ----------
+        """
+        import datetime
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:115]: I used it to check how long was taking to build the model, forgot to remove the import
done

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 14]
```diff
@@ -0,0 +1,1118 @@
...
+        self.lateral_contact_data = []
+        self.vertical_contact_data = []
+        # grid for simulation
+        self.simulation_grid = None
+        # parameters for each simulation and container for simulated values within the model
+        self.simulation_s_parameters = None  # input parameters for simulation
+        self.simulation_s_variogram = None  # inout variogram
+        self.simulation_sdata = (
+            None  # list [data of all sides, data of side S<0, data of side S>0]
+        )
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:60]: fixed, instance was not being used.

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 15]
```diff
@@ -0,0 +1,1118 @@
...
+        zmin, zmax : The minimum and maximum allowable data values. These are used in the back-transformation procedure.
+        nxdis, nydis : Number of discretization points for a block. If both nxdis and nydis are set to 1, then point kriging is performed.
+        ndmin, ndmax : The minimum and maximum number of original data that should be used to simulate a grid node. If there are fewer than ndmin data points, the node is not simulated.
+        radius : The maximum isotropic search radius.
+
+        Returns
+        s_parameters : dictionary of parameters to be used for lateral extent simulation. Parameters definition above
+        ----
+        """
+        if "tmin" in lateral_simulation_parameters:
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:255]: done

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 16]
```diff
@@ -0,0 +1,1118 @@
...
+        nug = 0; nst = 1                   # nugget effect = 0 and 1 nested structure
+        it1 = 1                            # nested structure type (1 - spherical, 2 - exponential, 3 - Gaussian)
+        cc1 = 1                            # One structure and no nugget
+        azi1 = 90                          # azimuth of this nested structure
+        hmaj1 = 9999 ; hmin1 = 9999   # range of this nested structure in the major (hmaj1) and minor (hmin1) direction
+
+        Returns
+        ----
+        """
+        if "nug" in lateral_simulation_parameters:
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:527]: done

File: LoopStructural/modelling/intrusions/intrusion_body.py [hunk 17]
```diff
@@ -0,0 +1,1118 @@
...
+            other_contact_data_xyz
+        )
+        other_contact_data.loc[:, "coord2"] = self.intrusion_frame[2].evaluate_value(
+            other_contact_data_xyz
+        )
+        other_contact_data.reset_index(inplace=True)
+
+        self.vertical_contact_data = [intrusion_network_data, other_contact_data]
+
+    def create_grid_for_simulation(self, spacing=None):
```
[the remark this answers is not in the document]
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:189]: done

File: LoopStructural/modelling/intrusions/intrusion_body.py (hunk 6 above)
Comment (Fer071989) [reply LoopStructural/modelling/intrusions/intrusion_body.py:340]: done

**Review by Fer071989 — APPROVED** [review]

**Pull request merged**